### ISSUE TITLE
Missing data in chunked SFTPClientFile read

### ISSUE THREAD

**Issue by dgvogol** [ronf/asyncssh#204]
Issue could be replicated using code snippet below:

```python
import logging
import asyncio
import asyncssh
import getpass
import os


async def sftp_issue():
    # create file with specified size
    file_name = os.path.expanduser('~/test.txt')
    file_size = 171*1024
    with open(file_name, 'wb') as fp:
        fp.write(os.urandom(file_size))

    logging.basicConfig(level=logging.DEBUG)
    logger = logging.getLogger()

    connection, client = await asyncssh.create_connection(None, 'localhost', 22,
                                                          username=getpass.getuser(), known_hosts=None,
                                                          client_keys=os.path.expanduser('~/.ssh/id_rsa'),
                                                          server_host_key_algs=[
                                                              'ecdsa-sha2-nistp521-cert-v01@openssh.com',
                                                              'ecdsa-sha2-nistp256-cert-v01@openssh.com',
                                                              'ecdsa-sha2-1.3.132.0.10-cert-v01@openssh.com',
                                                              'ssh-rsa-cert-v01@openssh.com',
                                                              'ssh-dss-cert-v01@openssh.com',
                                                              'ecdsa-sha2-nistp521',
                                                              'ecdsa-sha2-nistp384',
                                                              'ecdsa-sha2-nistp256',
                                                              'ecdsa-sha2-1.3.132.0.10',
                                                              'rsa-sha2-256',
                                                              'rsa-sha2-512',
                                                              'ssh-rsa',
                                                              'ssh-dss'])

    received_bytes = 0
    async with connection.start_sftp_client() as sftp:
        async with sftp.open(file_name, 'rb', encoding=None) as sftp_file:
            while True:
                data = await sftp_file.read(1024*64)
                if data:
                    received_bytes += len(data)
                    logger.info("total received bytes so far: 0x%08x", received_bytes)
                else:
                    break

    if received_bytes != file_size:
        logger.error('file size mismatch: expecting 0x%08x bytes, received 0x%08x bytes',
                     file_size, received_bytes)


if __name__ == '__main__':
    asyncio.get_event_loop().run_until_complete(sftp_issue())
```

Here is the output:
```
INFO:asyncssh:Opening SSH connection to localhost, port 22
INFO:asyncssh:[conn=0] Connection to localhost, port 22 succeeded
INFO:asyncssh:[conn=0]   Local address: 127.0.0.1, port 63918
DEBUG:asyncssh:[conn=0] Requesting key exchange
DEBUG:asyncssh:[conn=0] Received key exchange request
DEBUG:asyncssh:[conn=0] Beginning key exchange
DEBUG:asyncssh:[conn=0] Completed key exchange
INFO:asyncssh:[conn=0] Beginning auth for user xxxx
DEBUG:asyncssh:[conn=0] Trying public key auth with rsa-sha2-256 key
DEBUG:asyncssh:[conn=0] Signing request with rsa-sha2-256 key
INFO:asyncssh:[conn=0] Auth for user xxxx succeeded
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=0] Received unknown global request: hostkeys-00@openssh.com
INFO:asyncssh:[conn=0, chan=0]   Subsystem: sftp
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending init, version=3
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received version=3, extensions:
DEBUG:asyncssh.sftp:[conn=0, chan=0]   posix-rename@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   statvfs@openssh.com: 2
DEBUG:asyncssh.sftp:[conn=0, chan=0]   fstatvfs@openssh.com: 2
DEBUG:asyncssh.sftp:[conn=0, chan=0]   hardlink@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0]   fsync@openssh.com: 1
DEBUG:asyncssh.sftp:[conn=0, chan=0] Adjusting for non-standard symlink implementation
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending open for /home/xxxx/test.txt, mode 0x01
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 0 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 16384 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 32768 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 49152 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
INFO:root:total received bytes so far: 0x00010000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 65536 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 81920 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 98304 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 114688 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
INFO:root:total received bytes so far: 0x00020000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 131072 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 147456 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 163840 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 180224 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 11264 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending close for handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
INFO:asyncssh:[conn=0, chan=0] Received exit status 0
INFO:asyncssh:[conn=0, chan=0] Received channel close
INFO:asyncssh.sftp:[conn=0, chan=0] SFTP client exited
INFO:asyncssh:[conn=0, chan=0] Closing channel
INFO:asyncssh:[conn=0, chan=0] Channel closed
ERROR:root:file size mismatch: expecting 0x0002ac00 bytes, received 0x00020000 bytes

```

The file is read in 64k chunks, the first 2 64k are okay, but the last chunk (less than 64k) is lost, never returned to user application.

Did a little bit study, it seems the problem is within `class _SFTPParallelIO`

```python
class _SFTPParallellIO:
    '''omitted'''
    @asyncio.coroutine
    def run(self):
        """Perform all file I/O and return result or exception"""

        try:
            yield from self.start()

            self._start_tasks()

            while self._pending:
                done, self._pending = yield from asyncio.wait(
                    self._pending, return_when=asyncio.FIRST_COMPLETED)

                exceptions = []

                for task in done:
                    exc = task.exception()

                    if exc:
                        exceptions.append(exc)

                if exceptions:
                    for task in self._pending:
                        task.cancel()

                    raise exceptions[0]

                self._start_tasks()

            return (yield from self.finish())
        finally:
            yield from self.cleanup()

```
The first 2 64k is returned to user from `return (yield from self.finish())`, but the last chunk (less than 64k) is lost because of the 'End of file' exception.

**Comment by ronf** [comment]
Thank you for the report!

From an initial look, I agree that it seems to be related to the _SFTPParallelIO class. In your example here, it is breaking up your 64 KB reads into four parallel 16 KB reads. Those are working correctly, getting back a partial buffer on the last set of reads:

```
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 131072 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 147456 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 163840 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 180224 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 11264 data bytes
```

However, as you said, the fourth read getting back an EOF is causing a problem. I think the code was counting on getting back an empty string on the read() call when at EOF rather than an SFTPError of FX_EOF, but there might be some cases where that's not happening or perhaps that behavior varies depending on the server. I'll take a closer look.

In the meantime, you can work around this by either reducing your read sizes to 16384 to avoid triggering the parallel I/O, or increase the parallel I/O block size to match your read size by passing in the "block_size" argument in the SFTP open call.

**Comment by dgvogol** [comment]
> 
> 
> Thank you for the report!
> 
> From an initial look, I agree that it seems to be related to the _SFTPParallelIO class. In your example here, it is breaking up your 64 KB reads into four parallel 16 KB reads. Those are working correctly, getting back a partial buffer on the last set of reads:
> 
> ```
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 131072 in handle 00000000
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 147456 in handle 00000000
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 163840 in handle 00000000
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 16384 bytes at offset 180224 in handle 00000000
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 16384 data bytes
> DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 11264 data bytes
> ```
> 
> However, as you said, the fourth read getting back an EOF is causing a problem. I think the code was counting on getting back an empty string on the read() call when at EOF rather than an SFTPError of FX_EOF, but there might be some cases where that's not happening or perhaps that behavior varies depending on the server. I'll take a closer look.
> 
> In the meantime, you can work around this by either reducing your read sizes to 16384 to avoid triggering the parallel I/O, or increase the parallel I/O block size to match your read size by passing in the "block_size" argument in the SFTP open call.

Thanks, the work around works perfectly 👍

**Comment by ronf** [comment]
It looks like the fix for this is pretty simple. Here's a patch that should work:

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 1af1a09..69a96c6 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -452,7 +452,7 @@ class _SFTPParallelIO:
                 for task in done:
                     exc = task.exception()
 
-                    if exc:
+                    if exc and exc.code != FX_EOF:
                         exceptions.append(exc)
 
                 if exceptions:
```

This change and a unit test which tests for this specific edge case is in commit 0999b95 in the "develop" branch and will be part of the next release.

Thanks again for the report and for providing a reproducible test case!

**Comment by ronf** [comment]
This fix is now available in 1.16.1.
